[i18nIgnore] [reference] new components page (#9465)
withastro/docs: src/content/docs/en/reference/components-reference.mdx

---

> title: Component Reference

Reviewer: It should be in plural, right?
```suggestion
title: Components Reference
```

Author: Maybe even "Built-in Components Reference", since it's not a reference for generic components?

i18nReady: true
---
import Since from '~/components/Since.astro';
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';
import ReadMore from '~/components/ReadMore.astro';

## Built-in Components

Astro includes several built-in components for you to use in your projects. All built-in components are available in `.astro` files via `import {} from 'astro:components';`.

> You can reference the `Props` of these components using the [`ComponentProp` type](/en/guides/typescript/#componentprops-type) utility.

Reviewer 2: ```suggestion
You can reference the `Props` of these components using the [`ComponentProps` type](/en/guides/typescript/#componentprops-type) utility.
```

### `<Code />`

> ```astro 'theme="dark-plus"' /wrap\b/ /(inline) \/>/

Reviewer 2: This has nothing to do with the split but as I was rereading I was wondering... Shouldn't `defaultColor` also be highlighted? (if the goal is to highlight the attributes cited in line 36)
````suggestion
```astro 'theme="dark-plus"' /wrap\b/ /(inline) 'defaultColor={false}' \/>/
````

Note: I'm not used to Expressive Code, so feel free to double-check that I'm using the correct syntax. I took over the one from `theme`...

:::note
This component **does not** inherit the settings from your [Shiki configuration](/en/guides/markdown-content/#shiki-configuration). You will have to set them using the component props.
:::

#### Transformers

<p><Since v="4.11.0" /></p>

[Shiki transformers](https://shiki.style/packages/transformers#shikijs-transformers) can optionally be applied to code by passing them in through the `transformers` property as an array. Since Astro v4.14.0, you can also provide a string for [Shiki's `meta` attribute](https://shiki.style/guide/transformers#meta) to pass options to transformers.

Note that `transformers` only applies classes and you must provide your own CSS rules to target the elements of your code block.

```astro title="src/pages/index.astro" {12-13}
---
import { transformerNotationFocus, transformerMetaHighlight } from '@shikijs/transformers'
import { Code } from 'astro:components'
const code = `const foo = 'hello'
const bar = ' world'
console.log(foo + bar) // [!code focus]
`
---
<Code
  code={code}
  lang="js"
  transformers={[transformerMetaHighlight()]}

>   meta="{1,3}" />
>   
>   <style is:global>
>     pre.has-focused .line:not(.focused) {
>       filter: blur(1px);
>     }
>   </style>

Reviewer 2: nit: The first list is a matter of taste I guess but, as for the rest, `<style>` should not be indented.
```suggestion
  meta="{1,3}"
/>
  
<style is:global>
  pre.has-focused .line:not(.focused) {
    filter: blur(1px);
  }
</style>
```

##### `pictureAttributes`

An object of attributes to be added to the `<picture>` tag. Use this property to apply attributes to the outer `<picture>` element itself. Attributes applied to the `<Picture />` component directly will apply to the inner `<img>` element, except for those used for image transformation.

### `<Content />`

A generic component used to render the content of a [content collection entry](/en/guides/content-collections/#what-are-content-collections).

First, query one or more entries using `getCollection()` or `getEntry()`. Then, the `entry.render()` function can return the `<Content />` component for use in a `.astro` file template.

```astro title="src/pages/render-example.astro" {4, 7}
---
import { getEntry } from 'astro:content';
const entry = await getEntry('blog', 'post-1');
const { Content } = await entry.render();
---
<p>Published on: {entry.data.published.toDateString()}</p>
<Content />
```

### `<ViewTransitions />`

<p><Since v="2.9.0" /></p>

Opt in to using view transitions on individual pages by importing and adding the `<ViewTransitions />` routing component to `<head>` on every desired page.

> ```astro title="src/pages/index.astro" ins={2,7}

Reviewer 2: nit: All the code examples on the page use highlighting rather than diff notation so I think this one should do the same.
````suggestion
```astro title="src/pages/index.astro" {2,7}
````
Was there a specific reason for using `ins`? I'm guessing it was a copy/paste from an experimental feature/changeset.

Author: I'm going to guess that the `ins` was used to emphasize the "add this to otherwise unchanged code, and you get view transitions." Probably an artefact of this having been written when the component was new, and so all the text was about "adding" view transitions. (I do remember at the time we were HEAVY on this language because we were emphasizing you didn't just "turn them on", you added them to every page you wanted them on, or in a broader component, like a layout.

We can consider now that this is just "reference" and not "omg, here's how you add this new thing!" 😄

Reviewer 2: Afterwards, it's not necessarily annoying... it was more a question of consistency. But this can also be explained by the sentence above (“Opt in”) which perhaps "pushes more to action" than other descriptions so it can stay like that! (as I said, I was nitpicking 😄 ).

---
title: Built-in Components Reference
i18nReady: true
---
import Since from '~/components/Since.astro';
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';
import ReadMore from '~/components/ReadMore.astro';

## Built-in Components

> Astro includes several built-in components for you to use in your projects. All built-in components are available in `.astro` files via `import {} from 'astro:components';`.

Reviewer 3: ```suggestion
Astro includes several built-in components for you to use in your projects. All built-in components are available in `.astro` files via `import {} from 'astro:components'`.
```
Not sure if the semi is necessary in an inline code example, but it's been a while!

Author: Agreed. Looks a little funky.

You can reference the `Props` of these components using the [`ComponentProps` type](/en/guides/typescript/#componentprops-type) utility.

### `<Code />`

> ```astro 'theme="dark-plus"' /wrap\b/ /(inline) 'defaultColor={false}' \/>/

Reviewer 2: Sorry it seems I misplaced `defaultColor` in my previous suggestion... So both `inline` and `defaultColor` highlighting are broken. 😓 
````suggestion
```astro 'theme="dark-plus"' /wrap\b/ /(inline) \/>/ 'defaultColor={false}'
````
I think it should be good now... I don't know why I put it in the regex! 👀 (I should have done the review after and not before going to bed 😅 )

See the [list of languages supported by Prism](https://prismjs.com/#supported-languages) where you can find a language’s corresponding alias. And, you can also display your Astro code blocks with `lang="astro"`!

### `<Image />`

```astro title="src/components/MyComponent.astro"
---
// import the Image component and the image
import { Image } from 'astro:assets';
import myImage from "../assets/my_image.png"; // Image is 1600x900
---

<!-- `alt` is mandatory on the Image component -->
<Image src={myImage} alt="A description of my image." />
```

```html
<!-- Output -->
<!-- Image is optimized, proper attributes are enforced -->
<img
  src="/_astro/my_image.hash.webp"
  width="1600"
  height="900"
  decoding="async"
  loading="lazy"
  alt="A description of my image."
/>
```
#### Properties

- src (required)
- alt (required)
- width and height (required for `public/` and remote images)
- format
- quality
- densities

> - widths

Reviewer 3: These Properties don't follow the same format as the properties for `<Picture />`. Perhaps we should make them the same? I prefer the way of the Picture component, however, it might be lengthy here.

Author: We do spell these out on the image page, and I think rather than overhaul here now we'll keep this for a future fix because:

- we do want to rework the image page, so we could swap the loooong list there for this one, but that will also mean checking links etc.
- we might even go the Starlight route and make a page for each component! Even if we don't, with these things being the only things on their own page, we can make some standardization choices across all the different components.

So since that's a BIG change, let's consider it for a future improvement of this page!

---
title: Built-in Components Reference
i18nReady: true
tableOfContents:
  minHeadingLevel: 2

>   maxHeadingLevel: 2

Author: Until we go through and structure this page and we decide what it will look like, temporary fix so that only the components show up in the TOC.

(The alternative would be to convert `###` to `<h3>`, which would also be fine! This way preserves our normal Markdown authoring.)

---
import Since from '~/components/Since.astro';
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro';
import ReadMore from '~/components/ReadMore.astro';

> Astro includes several built-in components for you to use in your projects. All built-in components are available in `.astro` files via `import {} from 'astro:components'`.

Author: The heading that was here was redundant (pulled directly from the other page where it was simply a page section.)

This means we don't need it here, and all the components can be `##`.

> Astro includes several built-in components for you to use in your projects. All built-in components are available in `.astro` files via `import {} from 'astro:components'`.

Author: ```suggestion
Astro includes several built-in components for you to use in your projects. All built-in components are available in `.astro` files and may require an import statement.
```
